### PR TITLE
Add ZitterDelay for the Smooth Zitter mode

### PR DESCRIPTION
**Pull request by gabrielvicenteYT**
Good for anti-cheats such as Intave and Polar, working on it (will likely fail on the first build and CzechHek will make his own version, but whatever)

### CODE REVIEW

**Comment by ghost** [comment]
I'd say implement it using the tick method. It's way more stable. Now, as you mentioned Polar, Scaffold cannot properly bypass its scaffold checks, at least when it comes to bridging straight. The reason for that is because it needs to behave like an actual godbridger. Maybe bypassing it by using zitter could work, but it's unreliable/unstable.

**Comment by be4dev** [comment]
> Good for anti-cheats such as Intave and Polar, working on it (will likely fail on the first build and CzechHek will make his own version, but whatever)

well, there seems to be a logical problem with your code: every tick you pick a random value from min to max value and do other stuffs. You should make it the way, that it chooses one delay and waits it and then pick another (look at attackDelay code in KillAura for example)

**Comment by gabrielvicenteYT** [comment]
> > Good for anti-cheats such as Intave and Polar, working on it (will likely fail on the first build and CzechHek will make his own version, but whatever)
> 
> well, there seems to be a logical problem with your code: every tick you pick a random value from min to max value and do other stuffs. You should make it the way, that it chooses one delay and waits it and then pick another (look at attackDelay code in KillAura for example)

Thanks for the advice! I'll look into it

**Comment by gabrielvicenteYT** [comment]
> I'd say implement it using the tick method. It's way more stable. Now, as you mentioned Polar, Scaffold cannot properly bypass its scaffold checks, at least when it comes to bridging straight. The reason for that is because it needs to behave like an actual godbridger. Maybe bypassing it by using zitter could work, but it's unreliable/unstable.

I'm also trying to fix the issue that Smooth Zitter is kind of clumsy rn, and makes you place blocks where you don't want to.

**Comment by gabrielvicenteYT** [comment]
UPDATE: added auto GH Actions artifacts, https://github.com/CCBlueX/LiquidBounce/actions/runs/5562574736

**Comment by gabrielvicenteYT** [comment]
> I'd say implement it using the tick method. It's way more stable. Now, as you mentioned Polar, Scaffold cannot properly bypass its scaffold checks, at least when it comes to bridging straight. The reason for that is because it needs to behave like an actual godbridger. Maybe bypassing it by using zitter could work, but it's unreliable/unstable.

I think there should be an option to help zitter not throw a run

**Comment by gabrielvicenteYT** [comment]
@1zun4 Merge this, I'll improve it once I figure out how to make it work better. This is already needed, and works stable enough to merge.

**Comment by be4dev** [comment]
> @1zun4 Merge this, I'll improve it once I figure out how to make it work better. This is already needed, and works stable enough to merge.

I guess you still need to make delay work properly

**Comment by ghost** [comment]
And make delay work by ticks, instead of MS. It's more stable.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -22,7 +22,12 @@ jobs:
           cache: "gradle"
       - name: Build
         run: ./gradlew setupCiWorkspace && ./gradlew build
-      - name: Upload artifact
+      - name: Upload artifact to GH Actions
+        uses: actions/upload-artifact@v2
+        with:
+          name: LiquidBounce
+          path: build/libs
```
Comment (1zun4) [inline .github/workflows/build.yml]: Remove

Comment (gabrielvicenteYT) [reply .github/workflows/build.yml]: > Remove

I have a smarter idea

File: src/main/java/net/ccbluex/liquidbounce/features/module/modules/world/Scaffold.kt [hunk 2]
```diff
@@ -148,6 +148,18 @@ object Scaffold : Module("Scaffold", ModuleCategory.WORLD, Keyboard.KEY_I) {
...
+
+        override fun isSupported() = zitterMode == "Smooth"
+    }
+    private val maxZitterDelay by maxZitterDelayValue
+
+    private val minZitterDelay by object : IntegerValue("MinZitterDelay", 50, 0..250) {
+        override fun onChange(oldValue: Int, newValue: Int) = newValue.coerceAtMost(maxZitterDelay)
+
+        override fun isSupported() = zitterMode == "Smooth" && !maxZitterDelayValue.isMinimal()
+    }
```
Comment (1zun4) [inline src/main/java/net/ccbluex/liquidbounce/features/module/modules/world/Scaffold.kt]: first creating the value object and then by?

Comment (CzechHek) [reply src/main/java/net/ccbluex/liquidbounce/features/module/modules/world/Scaffold.kt]: to be able to use .isMinimal()

**Comment by gabrielvicenteYT** [comment]
> And make delay work by ticks, instead of MS. It's more stable.

Can't anti-cheats detect that, though?

**Comment by gabrielvicenteYT** [comment]
@1zun4 Removed what you requested to remove, merge now.

**Comment by ghost** [comment]
@gabrielvicenteYT, here: https://hastebin.com/share/ovasubaqin.kotlin

Since you can't use an IDE, you can copy and paste (assuming you can do that properly) the whole code from the link that I gave you. To keep it short and simple, this approach is more stable as the elapsed ticks' value (ticksSinceZitter variable) increases whenever the event (onUpdate) gets called (every tick).

**Comment by gabrielvicenteYT** [comment]
> @gabrielvicenteYT, here: https://hastebin.com/share/ovasubaqin.kotlin
> 
> Since you can't use an IDE, you can copy and paste (assuming you can do that properly) the whole code from the link that I gave you. To keep it short and simple, this approach is more stable as the elapsed ticks' value (ticksSinceZitter variable) increases whenever the event (onUpdate) gets called (every tick).

I am using IntelliJ right now, it's pretty decent, but the GUI does seem a bit bloated

**Comment by ghost** [comment]
> I am using IntelliJ right now, it's pretty decent, but the GUI does seem a bit bloated

PrayingHandsPrayGIF

**Comment by gabrielvicenteYT** [comment]
> > I am using IntelliJ right now, it's pretty decent, but the GUI does seem a bit bloated
> 
> PrayingHandsPrayGIF

How do I remove certain elements in IntelliJ's UI?

**Comment by gabrielvicenteYT** [comment]
@mems01 Yo, can you help me add some things to the Scaffold?

**Comment by gabrielvicenteYT** [comment]
Useless due to my new pull request, merge it when you can.

**Pull request closed without merging**